File: src/com/nubari/crackingthecodinginterview/java/StalinSort.java
package com.nubari.crackingthecodinginterview;

import java.util.ArrayList;

public class StalinSort {
    public static void main(String[] args) {
        var x = new int[]{4, 17, 5, 6, 12, 1, 9};
        var y = new ArrayList<Integer>();
        y.add(x[0]);
        for (int i = 1; i < x.length; i++) {
            if (x[i] > x[i - 1]) {
                y.add(x[i]);
            } else {
                var resultSize = y.size();
                y.remove(Integer.valueOf(x[i - 1]));
                if (resultSize != 0) {
                    if (y.size() == 0) {
                        y.add(x[i]);
                    } else if (x[i] > y.get(y.size() - 1)) {
                        y.add(x[i]);
                    }
                }

            }
        }
        System.out.println(y);
    }
}

File: src/com/nubari/crackingthecodinginterview/kotlin/StalinSort.kt
package com.nubari.crackingthecodinginterview


fun main() {
    println("Hello world!")
    val x: IntArray = intArrayOf(4, 17, 5, 6, 12, 1, 9);
    val result = stalinSort(x);
    println(result)

}

fun stalinSort(unsortedArray: IntArray): ArrayList<Int> {
    val stalinSorted = arrayListOf<Int>()
    stalinSorted.add(unsortedArray[0])
    for (i in unsortedArray.indices) {
        if (unsortedArray[i] > unsortedArray[i - 1]) {
            stalinSorted.add(unsortedArray[i])
        } else {
            val resultSize = stalinSorted.size
            stalinSorted.remove(unsortedArray[i - 1])
            if (resultSize != 0) {
                if (stalinSorted.size == 0) {
                    stalinSorted.add(unsortedArray[i])
                } else if (unsortedArray[i] > stalinSorted[stalinSorted.size - 1]) {
                    stalinSorted.add(unsortedArray[i])
                }
            }
        }
    }
    return stalinSorted
}